pkg/controller/application/app_controller.go

```diff
@@ -0,0 +1,295 @@
+		<-timeout.C
+		timeoutCh <- struct{}{}
+	}()
+	if ok := cache.WaitForCacheSync(timeoutCh, c.informer.HasSynced); !ok {
+		log.Logger.Error("Timeout expired during waiting for caches to sync. Skip starting the application controller")
+	} else {
+		log.Logger.Info("Starting application controller.")
+	}
+}
+
+/*
+Update the application in the scheduler
+*/
+func (appMgr *AppManager) updateApp(oldObj interface{}, newObj interface{}) {
+	newApp, ok := newObj.(*appv1.Application)
+	if !ok {
+		log.Logger.Error("newObj is not an Application")
+	}
+	oldApp, ok := oldObj.(*appv1.Application)
+	if !ok {
+		log.Logger.Error("oldObj is not an Application")
+	}
+	// No need to update if ResourceVersion is not changed
+	if newApp.ResourceVersion == oldApp.ResourceVersion {
+		log.Logger.Info("No need to update because app object is not modified")
+		return
+	}
+
+	// handle spec change
+	if !equality.Semantic.DeepEqual(oldApp.Spec, newApp.Spec) {
+		//TODO: handle update, discuss what should be updated since only the queue is the common property yet
+	}
+}
+
+/*
+Remove the application from the scheduler as well
+*/
+func (appMgr *AppManager) deleteApp(obj interface{}) {
+	app, ok := obj.(*appv1.Application)
+	if !ok {
+		log.Logger.Error("obj is not an Application")
+	}
+	appID := app.Name
+	err := appMgr.amProtocol.RemoveApplication(appID)
+	if err != nil {
+		log.Logger.Error("Application removal failed",
+			zap.String("appID", appID),
+			zap.Error(err))
+	}
+	log.Logger.Debug("App CRD deleted",
+		zap.String("Name", appID))
+	//TODO: delete related pods and deployments
+}
+
+/*
+Add application to scheduler
+*/
+func (appMgr *AppManager) addApp(obj interface{}) {
+	appCRD, ok := obj.(*appv1.Application)
+	if !ok {
+		log.Logger.Error("obj is not an Application")
+	}
+	if appMeta, ok := appMgr.getAppMetadata(appCRD); ok {
+		app := appMgr.amProtocol.GetApplication(appMeta.ApplicationID)
+		if app == nil {
+			appMgr.amProtocol.AddApplication(&interfaces.AddApplicationRequest{
+				Metadata: appMeta,
+				//TODO: debug recovery
+				//TODO: try to define the owner reference for the submitted pods to this CRD and try of we delete the
+				//CRD the pods will be deleted as well
+				Recovery: false,
+			})
+			// set and save status = New
+			appMgr.updateAppCRDStatus(appCRD, appv1.NewApplicationState)
+		}
+	}
+}
+
+func (appMgr *AppManager) updateAppCRDStatus(appCRD *appv1.Application, status appv1.ApplicationStateType) {
+	copy := appCRD.DeepCopy()
+	copy.Status = appv1.ApplicationStatus{
+		AppStatus:  status,
+		Message:    "app CRD status change",
+		LastUpdate: v1.NewTime(time.Now()),
+	}
+	_, err := appMgr.controller.appClient.ApacheV1alpha1().Applications(appCRD.Namespace).UpdateStatus(copy)
+	if err != nil {
+		log.Logger.Error("Failed to update application CRD",
+			zap.String("AppId", copy.Name))
+		return
+	}
+}
+
+func (appMgr *AppManager) getAppMetadata(app *appv1.Application) (interfaces.ApplicationMetadata, bool) {
+	appID := app.Name
```

Comment:
Like I commented here: https://github.com/apache/incubator-yunikorn-k8shim/pull/152#discussion_r453930011.
> According to https://kubernetes.io/docs/concepts/overview/working-with-objects/names/. Object UID should be unique across the cluster. Maybe we can use that as the applicationID. That means the user only needs to specify a namespace-unique name for each app-CRD. Also, metadata has generateName field, so if the user doesn't specify any name, K8s will generate one with that as the prefix. 

I think we should let `appID` = spec.UID, not the name, to ensure it is unique.
We can put the appName in the tag, in the format of `namespace/appName`

Reply:
Could you please fix this based on the comment here: https://issues.apache.org/jira/browse/YUNIKORN-340?focusedCommentId=17169061&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17169061

Reply:
Since using UID is more complicated, I will use namespace-name as applicationID

Reply:
Make sense to me, thanks